Add new functions

MaiconSoft opened the pull request:
Add UnixTime property;
Add Parse string as custom string;
Add ParseLocal to parse using local string like 'en-US';
All functions has test implemented.

MaiconSoft committed:
Add UnixTime and TotalSecounds

UnixTime:  Bidirectional property;
TotalSecounds: Bidirectional property;

MaiconSoft committed:
Add Parse and ParseLocal

- Parse a string using a format string;
- Parse a string using a local string;
- Tests add to unit.

colinj commented on DateTimeHelper.pas:
```diff
@@ -48,32 +48,31 @@   TDateTimeHelper = record helper for TDateTime
     class function GetTomorrow: TDateTime; static; inline;
     class function GetYesterDay: TDateTime; static; inline;
+    function GetUnixTime: Int64;
+    procedure SetUnixTime(const Value: Int64);
```
As stated in the 'README.md` file:

> An important rule ... is that TDateTime is a value type and therefore is **immutable**. This means all helper methods will create and **return a new value** and **not change the current value** of the calling TDateTime variable (unless explicitly assigned).

The philosophy of the TDateTimeHelper is that we don't modify the existing value but create a brand new value by using a class function.

Instead of the setter procedure `SetUnixTime`, turn it into a public class function (see class function Create). You could possibly call it class function `CreateUnixTime(const Value: Int64): TDateTime; static; inline;`.

colinj commented on DateTimeHelper.pas:
```diff
@@ -48,32 +48,31 @@   TDateTimeHelper = record helper for TDateTime
+    function GetUnixTime: Int64;
+    procedure SetUnixTime(const Value: Int64);
+    function GetTotalSecounds: Int64;
+    procedure SetTotalSecounds(const Value: Int64);
```
Same as `SetUnixTime`, make `SetTotalSecounds` a class function in the public section. Make sure you spell `seconds` correctly. The class function can be `CreateTotalSeconds(const Value: Int64): TDateTime; static; inline;`

colinj commented on DateTimeHelper.pas:
```diff
@@ -108,15 +105,35 @@   TDateTimeHelper = record helper for TDateTime
+      Int64): Boolean; inline;
+
+    // Parse a string as TDateTime, using a string format ex.('MM/dd/yyyy hh:mm:ss')
+    function Parse(Date: string; aFormat: string = ''; aDateSeparator: Char = #0;
```
Make `Parse` a class function similar to `class function Create`. This function should always return a new value and not modify the instance's value.

colinj commented on DateTimeHelper.pas:
```diff
@@ -108,15 +105,35 @@   TDateTimeHelper = record helper for TDateTime
+      aTimeSeparator: Char = #0):TDateTime; inline;
+
+   // Parse a string as TDateTime, using local string ex. ('en-US')
+    function ParseLocal(Date: string;local:string=''):TDateTime; inline;
```
Treat `ParseLocal` as the same as `Parse`. Make this a class function.

colinj commented on DateTimeHelper.pas:
```diff
@@ -108,15 +105,35 @@   TDateTimeHelper = record helper for TDateTime
+   // Parse a string as TDateTime, using local string ex. ('en-US')
+    function ParseLocal(Date: string;local:string=''):TDateTime; inline;
+
+    property UnixTime: Int64 read GetUnixTime write SetUnixTime;
```
`TDateTimeHelper` properties can only be read-only. Please remove the setter `write SetUnixTime`.

colinj commented on DateTimeHelper.pas:
```diff
@@ -108,15 +105,35 @@   TDateTimeHelper = record helper for TDateTime
+    property UnixTime: Int64 read GetUnixTime write SetUnixTime;
+
+    // Total Secounds until initial Date (12/30/1899 12:00 am)
+    property TotalSecounds: Int64 read GetTotalSecounds write SetTotalSecounds;
```
The spelling should be `TotalSeconds` and please remove the setter.

colinj commented on DateTimeHelper.pas:
```diff
@@ -108,15 +105,35 @@   TDateTimeHelper = record helper for TDateTime
     function MinutesBetween(const aDateTime: TDateTime): Int64; inline;
     function SecondsBetween(const aDateTime: TDateTime): Int64; inline;
     function MilliSecondsBetween(const aDateTime: TDateTime): Int64; inline;
+    function WithinYears(const aDateTime: TDateTime; const aYears: Integer):
```
Avoid changing the formatting of these lines since this is not part of the PR

colinj commented on DateTimeHelper.pas:
```diff
@@ -82,24 +81,22 @@   TDateTimeHelper = record helper for TDateTime
     function Equals(const aDateTime: TDateTime): Boolean; inline;
     function IsSameDay(const aDateTime: TDateTime): Boolean; inline;
-    function InRange(const aStartDateTime, aEndDateTime: TDateTime; const aInclusive: Boolean = True): Boolean; inline;
+    function InRange(const aStartDateTime, aEndDateTime: TDateTime; const
```
Please don't change the formatting. This is not part of the PR.

colinj commented on DateTimeHelper.pas:
```diff
@@ -48,32 +48,31 @@   TDateTimeHelper = record helper for TDateTime
+    procedure SetTotalSecounds(const Value: Int64);
   public
-    class function Create(const aYear, aMonth, aDay: Word): TDateTime; overload; static; inline;
+    class function Create(const aYear, aMonth, aDay: Word): TDateTime; overload;
```
Please don't change the formatting. This is not part of the PR.

colinj commented on DateTimeHelper.pas:
```diff
@@ -108,15 +105,35 @@   TDateTimeHelper = record helper for TDateTime
+    // Total Secounds until initial Date (12/30/1899 12:00 am)
+    property TotalSecounds: Int64 read GetTotalSecounds write SetTotalSecounds;
 
-    function WithinYears(const aDateTime: TDateTime; const aYears: Integer): Boolean; inline;
```
Please keep this formatting unless you are making real changes to these functions.

colinj commented on DateTimeHelper.pas:
```diff
@@ -270,6 +286,16 @@ class function TDateTimeHelper.GetTomorrow: TDateTime;
 
+function TDateTimeHelper.GetTotalSecounds: Int64;
+begin
+  Result := Trunc(86400 * self);
```
Rename `self` to `Self` to keep formatting standards. Please make sure all instances of 'self' is also renamed to 'Self'

colinj commented on DateTimeHelper.pas:
```diff
@@ -285,7 +311,8 @@ function TDateTimeHelper.HoursBetween(const aDateTime: TDateTime): Int64;
 end;
 
-function TDateTimeHelper.InRange(const aStartDateTime, aEndDateTime: TDateTime; const aInclusive: Boolean): Boolean;
+function TDateTimeHelper.InRange(const aStartDateTime, aEndDateTime: TDateTime;
```
Please don't change the formatting. This is not part of the PR.

colinj commented on DateTimeHelper.pas:
```diff
@@ -335,6 +362,16 @@ function TDateTimeHelper.SecondsBetween(const aDateTime: TDateTime): Int64;
   Result := System.DateUtils.SecondsBetween(Self, aDateTime);
 end;
 
+procedure TDateTimeHelper.SetTotalSecounds(const Value: Int64);
```
As mentioned above, we don't modify the value of existing instance. Instead this setter function needs to be turned into a class function that returns new value. 

```
class function CreateTotalSeconds(const Value: Int64);
begin
  Result := (Value / 86400);
end;
```

colinj commented on DateTimeHelper.pas:
```diff
@@ -335,6 +362,16 @@ function TDateTimeHelper.SecondsBetween(const aDateTime: TDateTime): Int64;
+  self := (Value / 86400);
+end;
+
+procedure TDateTimeHelper.SetUnixTime(const Value: Int64);
```
Replace setter procedure into a class function and change it from `Self` to `Result`.

colinj commented on DateTimeHelper.pas:
```diff
@@ -421,4 +458,59 @@ function TDateTimeHelper.YearsBetween(const aDateTime: TDateTime): Integer;
   Result := System.DateUtils.YearsBetween(Self, aDateTime);
 end;
 
+function TDateTimeHelper.Parse(Date: string; aFormat: string = ''; aDateSeparator: Char = #0;
```
Make this into a class function and don't use `Date` as an argument as it can be ambiguous to the Delphi `Date` function. Rename it as 'aDate' where the prefix `a` stands for "argument".

colinj commented on DateTimeHelper.pas:
```diff
@@ -421,4 +458,59 @@ function TDateTimeHelper.YearsBetween(const aDateTime: TDateTime): Integer;
+
+  if aLength = 0 then
+  begin
+    self := StrToDateTime(Date);
```
Do not update `Self`. Just assign value to `Result`.

colinj commented on DateTimeHelper.pas:
```diff
@@ -421,4 +458,59 @@ function TDateTimeHelper.YearsBetween(const aDateTime: TDateTime): Integer;
+        ShortTimeFormat := aFormats[1];
+    end;
+  end;
+  Self := StrToDateTime(Date, fs);
```
Do not assign to `Self` but return value via `Result`

colinj commented on DateTimeHelper.pas:
```diff
@@ -421,4 +458,59 @@ function TDateTimeHelper.YearsBetween(const aDateTime: TDateTime): Integer;
+    result:= self;
+end;
+
+function TDateTimeHelper.ParseLocal(Date: string; local: string = ''):TDateTime;
```
Treat 'ParseLocal' the same  as `Parse`.

colinj commented on DateTimeHelper.pas:
```diff
@@ -421,4 +458,59 @@ function TDateTimeHelper.YearsBetween(const aDateTime: TDateTime): Integer;
+  else
+    fs := TFormatSettings.Create(local);
+
+  Self := StrToDateTime(Date, fs);
```
Don't set 'Self'. Only set `Result`

colinj reviewed (changes requested):
Thank you @MaiconSoft for making this PR for adding `UnixTime` and `TotalSeconds`. Unfortunately, I can't accept these changes as they are at the moment because it would break the principle of mutating a value type. I've provided comments explaining what is required changing to follow the principle.

Also could you please remove the changes to the lines of code which are not part of this PR.
I hope you consider making those changes as I would really like to accept this PR eventually.

MaiconSoft committed:
Fix not immutable functions

- Change SetUnixTime to CreateUnixtime;
- Change SetTotalSecounds to CreateTotalSecounds;
- Change Parse and ParseLocal to class function;
- Change TotalSecounds and Unixtime to readonly property;
- Added overload function to Create, for construct a TDateTime with string;
- Added CreateLocal for create using parseLocal;
- Fix Test unit for acept changes above.

MaiconSoft commented:
Hi Colinj, sorry about that, i not realize this rule, but i am open for all changes needed for help your project. Thank you

colinj commented:
Thank you @MaiconSoft 
There were still a few spelling/formatting issues but generally speaking the PR was fine.
